src/wal/src/local_storage_impl/segment.rs

```diff
@@ -213,154 +238,270 @@ impl Segment {
 
         // Validate segment header
         let header_len = SEGMENT_HEADER.len();
-        ensure!(size >= header_len as u64, InvalidHeader);
+        ensure!(
+            file_size >= (VERSION_SIZE + header_len) as u64,
+            InvalidHeader
+        );
         let header = &mmap[VERSION_SIZE..VERSION_SIZE + header_len];
         ensure!(header == SEGMENT_HEADER, InvalidHeader);
 
         // Read and validate all records
         let mut pos = VERSION_SIZE + header_len;
         let mut record_position = Vec::new();
-        while pos < size as usize {
-            let data = &mmap[pos..];
 
-            let record = self
-                .record_encoding
-                .decode(data)
-                .box_err()
-                .context(InvalidRecord)?;
+        // Update min and max sequence number
+        let mut min_seq = MAX_SEQUENCE_NUMBER;
+        let mut max_seq = MIN_SEQUENCE_NUMBER;
 
-            record_position.push(Position {
-                start: pos as u64,
-                end: (pos + record.len()) as u64,
-            });
+        while pos < file_size as usize {
+            let data = &mmap[pos..];
 
-            // Move to the next record
-            pos += record.len();
+            match self.record_encoding.decode(data).box_err() {
+                Ok(record) => {
+                    record_position.push(Position {
+                        start: pos,
+                        end: pos + record.len(),
+                    });
+                    min_seq = min_seq.min(record.sequence_num);
+                    max_seq = max_seq.max(record.sequence_num);
+                    pos += record.len();
+                }
+                Err(_) => {
+                    // If decoding fails, we've reached the end of valid data
```

Comment:
This is too tricky, maybe we can add some header to each record, and check it before decode.

You can leave a TODO comment here, we can refactor it later.